velox/exec/HashBuild.cpp

```diff
@@ -102,17 +118,59 @@ void HashBuild::setupTable() {
         (joinNode_->isLeftSemiJoin() || joinNode_->isNullAwareAntiJoin());
     // Right semi join needs to tag build rows that were probed.
     const bool needProbedFlag = joinNode_->isRightSemiJoin();
-    // Ignore null keys
-    table_ = HashTable<true>::createForJoin(
-        std::move(keyHashers),
-        dependentTypes,
-        !dropDuplicates, // allowDuplicates
-        needProbedFlag, // hasProbedFlag
-        mappedMemory_);
+    if (joinNode_->isNullAwareAntiJoin() && joinNode_->filter()) {
+      // We need to check null key rows in build side in this case.
+      table_ = HashTable<false>::createForJoin(
+          std::move(keyHashers),
+          dependentTypes,
+          !dropDuplicates, // allowDuplicates
+          needProbedFlag, // hasProbedFlag
+          mappedMemory_);
+    } else {
+      // Ignore null keys
+      table_ = HashTable<true>::createForJoin(
+          std::move(keyHashers),
+          dependentTypes,
+          !dropDuplicates, // allowDuplicates
+          needProbedFlag, // hasProbedFlag
+          mappedMemory_);
+    }
   }
   analyzeKeys_ = table_->hashMode() != BaseHashTable::HashMode::kHash;
 }
 
+void HashBuild::removeFilterInputNullRows() {
+  bool changed = false;
+  VELOX_DCHECK(std::is_sorted(keyChannels_.begin(), keyChannels_.end()));
```

Comment:
Why would that be the case all the time?

Reply:
Good catch, I think I only checked `dependentChannels_` and missed `keyChannels_`